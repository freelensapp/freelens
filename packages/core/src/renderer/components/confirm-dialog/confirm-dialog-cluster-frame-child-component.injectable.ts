/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { clusterFrameChildComponentInjectionToken } from "@freelensapp/react-application";
import { getInjectable } from "@ogre-tools/injectable";
import { computed } from "mobx";
import { ConfirmDialog } from "./confirm-dialog";

const confirmDialogClusterFrameChildComponentInjectable = getInjectable({
  id: "confirm-dialog-cluster-frame-child-component",

  instantiate: () => ({
    id: "confirm-dialog",
    shouldRender: computed(() => true),
    Component: ConfirmDialog,
  }),

  injectionToken: clusterFrameChildComponentInjectionToken,
});

export default confirmDialogClusterFrameChildComponentInjectable;
