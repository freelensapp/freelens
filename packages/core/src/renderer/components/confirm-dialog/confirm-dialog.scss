/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

@use "../vars" as *;

.ConfirmDialog {
  $spacing: $padding * 3;

  &.warning {
    .Button.ok {
      background: var(--colorSoftError);
    }
  }

  > .box {
    max-width: 50vw;
    min-width: 45 * $unit;
    background-color: white;
    border-radius: $radius;
    line-height: 1.5;
  }

  .confirm-content {
    padding: $spacing;
    display: flex;
    align-items: center;
    word-break: break-word;

    > .Icon {
      margin-left: inherit;
      margin-right: $margin * 2;
      color: var(--colorSoftError);
    }

    > :not(.Icon) {
      flex: 1;
    }

    .warning {
      font-size: small;
      margin-top: $margin;
    }
  }

  .confirm-buttons {
    background: #f4f4f4;
    padding: $padding * 2.5;
    display: flex;
    justify-content: flex-end;
    border-bottom-left-radius: $radius;
    border-bottom-right-radius: $radius;

    > * {
      margin-left: $margin
    }
  }
}
