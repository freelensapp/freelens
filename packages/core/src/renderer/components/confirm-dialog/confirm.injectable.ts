/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { getInjectable } from "@ogre-tools/injectable";
import type { ConfirmDialogBooleanParams } from "./confirm-dialog";
import openConfirmDialogInjectable from "./open.injectable";

export type Confirm = (params: ConfirmDialogBooleanParams) => Promise<boolean>;

const confirmInjectable = getInjectable({
  id: "confirm",
  instantiate: (di): Confirm => {
    const open = di.inject(openConfirmDialogInjectable);

    return (params) =>
      new Promise((resolve) => {
        open({
          ok: () => resolve(true),
          cancel: () => resolve(false),
          ...params,
        });
      });
  },
});

export default confirmInjectable;
