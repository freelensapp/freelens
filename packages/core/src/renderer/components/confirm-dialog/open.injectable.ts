/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { getInjectable } from "@ogre-tools/injectable";
import type { ConfirmDialogParams } from "./confirm-dialog";
import confirmDialogStateInjectable from "./state.injectable";

export type OpenConfirmDialog = (params: ConfirmDialogParams) => void;

const openConfirmDialogInjectable = getInjectable({
  id: "open-confirm-dialog",
  instantiate: (di): OpenConfirmDialog => {
    const state = di.inject(confirmDialogStateInjectable);

    return (params) => state.set(params);
  },
});

export default openConfirmDialogInjectable;
