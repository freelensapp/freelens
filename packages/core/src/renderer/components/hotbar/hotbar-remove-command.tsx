/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { withInjectables } from "@ogre-tools/injectable-react";
import type { IComputedValue } from "mobx";
import { observer } from "mobx-react";
import React from "react";
import type { ComputeHotbarDisplayLabel } from "../../../features/hotbar/storage/common/compute-display-label.injectable";
import computeHotbarDisplayLabelInjectable from "../../../features/hotbar/storage/common/compute-display-label.injectable";
import type { Hotbar } from "../../../features/hotbar/storage/common/hotbar";
import hotbarsInjectable from "../../../features/hotbar/storage/common/hotbars.injectable";
import type { RemoveHotbar } from "../../../features/hotbar/storage/common/remove.injectable";
import removeHotbarInjectable from "../../../features/hotbar/storage/common/remove.injectable";
import commandOverlayInjectable from "../command-palette/command-overlay.injectable";
import type { OpenConfirmDialog } from "../confirm-dialog/open.injectable";
import openConfirmDialogInjectable from "../confirm-dialog/open.injectable";
import { Select } from "../select";

interface Dependencies {
  closeCommandOverlay: () => void;
  openConfirmDialog: OpenConfirmDialog;
  hotbars: IComputedValue<Hotbar[]>;
  computeHotbarDisplayLabel: ComputeHotbarDisplayLabel;
  removeHotbar: RemoveHotbar;
}

const NonInjectedHotbarRemoveCommand = observer(
  ({ closeCommandOverlay, openConfirmDialog, hotbars, computeHotbarDisplayLabel, removeHotbar }: Dependencies) => (
    <Select
      menuPortalTarget={null}
      onChange={(option) => {
        if (!option) {
          return;
        }

        closeCommandOverlay();
        openConfirmDialog({
          okButtonProps: {
            label: "Remove Hotbar",
            primary: false,
            accent: true,
          },
          ok: () => removeHotbar(option.value),
          message: (
            <div className="confirm flex column gaps">
              <p>
                Are you sure you want remove hotbar <b>{option.value.name.get()}</b>?
              </p>
            </div>
          ),
        });
      }}
      components={{ DropdownIndicator: null, IndicatorSeparator: null }}
      menuIsOpen={true}
      options={hotbars.get().map((hotbar) => ({
        value: hotbar,
        label: computeHotbarDisplayLabel(hotbar),
      }))}
      autoFocus={true}
      escapeClearsValue={false}
      placeholder="Remove hotbar"
    />
  ),
);

export const HotbarRemoveCommand = withInjectables<Dependencies>(NonInjectedHotbarRemoveCommand, {
  getProps: (di, props) => ({
    ...props,
    closeCommandOverlay: di.inject(commandOverlayInjectable).close,
    openConfirmDialog: di.inject(openConfirmDialogInjectable),
    computeHotbarDisplayLabel: di.inject(computeHotbarDisplayLabelInjectable),
    hotbars: di.inject(hotbarsInjectable),
    removeHotbar: di.inject(removeHotbarInjectable),
  }),
});
