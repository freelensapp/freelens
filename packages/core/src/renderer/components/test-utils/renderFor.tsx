/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { DiContainer } from "@ogre-tools/injectable";
import { DiContextProvider } from "@ogre-tools/injectable-react";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";

export type DiRender = (ui: React.ReactElement) => RenderResult;

type DiRenderFor = (di: DiContainer) => DiRender;

export const renderFor: DiRenderFor = (di) => (ui) => {
  const result = testingLibraryRender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>);

  return {
    ...result,

    rerender: (ui: React.ReactElement) => result.rerender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>),
  };
};
