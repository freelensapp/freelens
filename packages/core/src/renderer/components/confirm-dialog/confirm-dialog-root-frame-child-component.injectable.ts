/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable } from "@ogre-tools/injectable";
import { rootFrameChildComponentInjectionToken } from "@freelens/react-application";
import { computed } from "mobx";
import { ConfirmDialog } from "./confirm-dialog";

const confirmDialogRootFrameChildComponentInjectable = getInjectable({
  id: "confirm-dialog-root-frame-child-component",

  instantiate: () => ({
    id: "confirm-dialog",
    shouldRender: computed(() => true),
    Component: ConfirmDialog,
  }),

  injectionToken: rootFrameChildComponentInjectionToken,
});

export default confirmDialogRootFrameChildComponentInjectable;
