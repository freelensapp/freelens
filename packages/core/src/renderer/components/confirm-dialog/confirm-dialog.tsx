/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./confirm-dialog.scss";

import type { ButtonProps } from "@freelensapp/button";
import { Button } from "@freelensapp/button";
import { Icon } from "@freelensapp/icon";
import type { ShowNotification } from "@freelensapp/notifications";
import { showErrorNotificationInjectable } from "@freelensapp/notifications";
import type { StrictReactNode } from "@freelensapp/utilities";
import { cssNames, noop, prevDefault } from "@freelensapp/utilities";
import { withInjectables } from "@ogre-tools/injectable-react";
import type { IObservableValue } from "mobx";
import { computed, makeObservable, observable } from "mobx";
import { observer } from "mobx-react";
import React from "react";
import type { DialogProps } from "../dialog";
import { Dialog } from "../dialog";
import confirmDialogStateInjectable from "./state.injectable";

export interface ConfirmDialogProps extends Partial<DialogProps> {}

export interface ConfirmDialogParams extends ConfirmDialogBooleanParams {
  ok?: () => any | Promise<any>;
  cancel?: () => any | Promise<any>;
}

export interface ConfirmDialogBooleanParams {
  labelOk?: StrictReactNode;
  labelCancel?: StrictReactNode;
  message: StrictReactNode;
  icon?: StrictReactNode;
  okButtonProps?: Partial<ButtonProps>;
  cancelButtonProps?: Partial<ButtonProps>;
}

interface Dependencies {
  state: IObservableValue<ConfirmDialogParams | undefined>;
  showErrorNotification: ShowNotification;
}

const defaultParams = {
  ok: noop,
  cancel: noop,
  labelOk: "Ok",
  labelCancel: "Cancel",
  icon: <Icon big material="warning" />,
};

@observer
class NonInjectedConfirmDialog extends React.Component<ConfirmDialogProps & Dependencies> {
  @observable isSaving = false;

  constructor(props: ConfirmDialogProps & Dependencies) {
    super(props);
    makeObservable(this);
  }

  @computed
  get params() {
    return Object.assign({}, defaultParams, this.props.state.get() ?? ({} as ConfirmDialogParams));
  }

  ok = async () => {
    try {
      this.isSaving = true;
      await (async () => this.params.ok())();
    } catch (error) {
      this.props.showErrorNotification(
        <>
          <p>Confirmation action failed:</p>
          <p>
            {error instanceof Error
              ? error.message
              : typeof error === "string"
                ? error
                : "Unknown error occurred while ok-ing"}
          </p>
        </>,
      );
    } finally {
      this.isSaving = false;
      this.props.state.set(undefined);
    }
  };

  onClose = () => {
    this.isSaving = false;
  };

  close = async () => {
    try {
      await Promise.resolve(this.params.cancel());
    } catch (error) {
      this.props.showErrorNotification(
        <>
          <p>Cancelling action failed:</p>
          <p>
            {error instanceof Error
              ? error.message
              : typeof error === "string"
                ? error
                : "Unknown error occurred while cancelling"}
          </p>
        </>,
      );
    } finally {
      this.isSaving = false;
      this.props.state.set(undefined);
    }
  };

  render() {
    const { state, className, ...dialogProps } = this.props;
    const isOpen = Boolean(state.get());
    const { icon, labelOk, labelCancel, message, okButtonProps = {}, cancelButtonProps = {} } = this.params;

    return (
      <Dialog
        {...dialogProps}
        className={cssNames("ConfirmDialog", className)}
        isOpen={isOpen}
        onClose={this.onClose}
        close={this.close}
        {...(isOpen ? { "data-testid": "confirmation-dialog" } : {})}
      >
        <div className="confirm-content">
          {icon} {message}
        </div>
        <div className="confirm-buttons">
          <Button
            plain
            className="cancel"
            label={labelCancel}
            onClick={prevDefault(this.close)}
            {...cancelButtonProps}
          />
          <Button
            autoFocus
            primary
            className="ok"
            label={labelOk}
            onClick={prevDefault(this.ok)}
            waiting={this.isSaving}
            data-testid="confirm"
            {...okButtonProps}
          />
        </div>
      </Dialog>
    );
  }
}

export const ConfirmDialog = withInjectables<Dependencies, ConfirmDialogProps>(NonInjectedConfirmDialog, {
  getProps: (di, props) => ({
    ...props,
    state: di.inject(confirmDialogStateInjectable),
    showErrorNotification: di.inject(showErrorNotificationInjectable),
  }),
});
