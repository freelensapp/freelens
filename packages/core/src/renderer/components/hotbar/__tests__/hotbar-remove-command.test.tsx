/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "@testing-library/jest-dom";
import type { RenderResult } from "@testing-library/react";
import { fireEvent } from "@testing-library/react";
import type { IComputedValue } from "mobx";
import React from "react";
import directoryForUserDataInjectable from "../../../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";
import type { CreateHotbar } from "../../../../features/hotbar/storage/common/create-hotbar.injectable";
import createHotbarInjectable from "../../../../features/hotbar/storage/common/create-hotbar.injectable";
import type { Hotbar } from "../../../../features/hotbar/storage/common/hotbar";
import hotbarsInjectable from "../../../../features/hotbar/storage/common/hotbars.injectable";
import hotbarsStateInjectable from "../../../../features/hotbar/storage/common/state.injectable";
import { getDiForUnitTesting } from "../../../getDiForUnitTesting";
import { ConfirmDialog } from "../../confirm-dialog";
import { renderFor } from "../../test-utils/renderFor";
import { HotbarRemoveCommand } from "../hotbar-remove-command";

describe("<HotbarRemoveCommand />", () => {
  let result: RenderResult;
  let createHotbar: CreateHotbar;
  let hotbars: IComputedValue<Hotbar[]>;

  beforeEach(() => {
    const di = getDiForUnitTesting();
    const render = renderFor(di);

    di.override(directoryForUserDataInjectable, () => "/some-directory-for-user-data");

    createHotbar = di.inject(createHotbarInjectable);
    hotbars = di.inject(hotbarsInjectable);

    const hotbarsState = di.inject(hotbarsStateInjectable);
    const defaultHotbar = createHotbar({ name: "default" });
    const nonDefaultHotbar = createHotbar({ name: "non-default" });

    hotbarsState.set(defaultHotbar.id, defaultHotbar);
    hotbarsState.set(nonDefaultHotbar.id, nonDefaultHotbar);

    result = render(
      <>
        <HotbarRemoveCommand />
        <ConfirmDialog />
      </>,
    );
  });

  it("renders w/o errors", () => {
    expect(result.container).toMatchSnapshot();
  });

  it("calls remove if you click on the entry", () => {
    fireEvent.click(result.getByText("1: default"));
    fireEvent.click(result.getByText("Remove Hotbar"));
    expect(hotbars.get().length).toBe(1);
  });
});
