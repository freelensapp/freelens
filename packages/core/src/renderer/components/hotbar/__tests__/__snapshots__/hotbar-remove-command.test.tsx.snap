// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HotbarRemoveCommand /> renders w/o errors 1`] = `
<div>
  <div
    class="Select theme-dark css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-2-live-region"
    >
      <span
        id="aria-selection"
      >
        option , selected.
      </span>
      <span
        id="aria-context"
      >
         2 results available. Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
      </span>
    </span>
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class="Select__control Select__control--is-focused Select__control--menu-is-open css-t3ipsp-control"
    >
      <div
        class="Select__value-container css-hlgwow"
      >
        <div
          class="Select__placeholder css-1jqq78o-placeholder"
          id="react-select-2-placeholder"
        >
          Remove hotbar
        </div>
        <div
          class="Select__input-container css-19bb58m"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-controls="react-select-2-listbox"
            aria-describedby="react-select-2-placeholder"
            aria-expanded="true"
            aria-haspopup="true"
            aria-owns="react-select-2-listbox"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class="Select__input"
            id="react-select-2-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="Select__indicators css-1wy0on6"
      />
    </div>
    <div
      class="theme-dark Select__menu css-1nmdiq5-menu"
      id="react-select-2-listbox"
    >
      <div
        class="Select__menu-list css-qr46ko"
      >
        <div
          aria-disabled="false"
          class="Select__option css-10wo9uf-option"
          id="react-select-2-option-0"
          tabindex="-1"
        >
          1: default
        </div>
        <div
          aria-disabled="false"
          class="Select__option css-10wo9uf-option"
          id="react-select-2-option-1"
          tabindex="-1"
        >
          2: non-default
        </div>
      </div>
    </div>
  </div>
</div>
`;
